File: basic_dsa_one_hundred_eighty_three.java
/**
 * Basic DSA Problems - Find Minimum in Rotated Sorted Array
 * Min element in sorted array rotated at unknown pivot (no duplicates)
 */
public class basic_dsa_one_hundred_eighty_three {

    // Problem: Sorted array rotated at unknown pivot. Find minimum. No duplicates.

    /**
     * Binary search: if nums[mid] > nums[hi], min is in (mid, hi]; else min is in [lo, mid].
     * Time O(log n), Space O(1)
     */
    public static int findMin(int[] nums) {
        int lo = 0, hi = nums.length - 1;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] > nums[hi]) lo = mid + 1;
            else hi = mid;
        }
        return nums[lo];
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Find Min in Rotated Sorted Array");
        System.out.println("=".repeat(60));

        System.out.println("[3,4,5,1,2] -> " + findMin(new int[]{3, 4, 5, 1, 2}));
        System.out.println("[4,5,6,7,0,1,2] -> " + findMin(new int[]{4, 5, 6, 7, 0, 1, 2}));
        System.out.println("[1] -> " + findMin(new int[]{1}));
    }
}

File: basic_dsa_one_hundred_eighty_three.py
"""
Basic DSA - Search Insert Position
Given a sorted array of distinct integers nums and a target value, return the index
if the target is found. If not, return the index where it would be inserted in order.
"""


def search_insert(nums, target):
    """
    Lower bound binary search (first index with nums[i] >= target).
    Time Complexity: O(log n)
    Space Complexity: O(1)
    """
    lo, hi = 0, len(nums)
    while lo < hi:
        mid = (lo + hi) // 2
        if nums[mid] < target:
            lo = mid + 1
        else:
            hi = mid
    return lo


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Search Insert Position")
    print("=" * 60)

    print(search_insert([1, 3, 5, 6], 5))  # 2
    print(search_insert([1, 3, 5, 6], 2))  # 1
    print(search_insert([1, 3, 5, 6], 7))  # 4
    print(search_insert([1, 3, 5, 6], 0))  # 0
